diff --git a/cpp/open3d/core/nns/kernel/L2Select.cuh b/cpp/open3d/core/nns/kernel/L2Select.cuh
index 222a191..31dc12d 100644
--- a/cpp/open3d/core/nns/kernel/L2Select.cuh
+++ b/cpp/open3d/core/nns/kernel/L2Select.cuh
@@ -59,8 +59,9 @@ __global__ void l2SelectMin1(T* productDistances,
                              int row_stride,
                              int out_row_stride) {
     // Each block handles kRowsPerBlock rows of the distances (results)
+    extern __shared__ unsigned char shared_mem[];
+    Pair<T, int>* blockMin = reinterpret_cast<Pair<T, int>*>(shared_mem);
     Pair<T, int> threadMin[kRowsPerBlock];
-    __shared__ Pair<T, int> blockMin[kRowsPerBlock * (kBlockSize / kWarpSize)];
 
     T distance[kRowsPerBlock];
 
@@ -220,6 +221,9 @@ void runL2SelectMin(const cudaStream_t stream,
         constexpr int kRowsPerBlock = 8;
 
         auto block = dim3(kThreadsPerBlock);
+        constexpr size_t kBlockMinCount =
+                kRowsPerBlock * (kThreadsPerBlock / kWarpSize);
+        size_t shared_mem_size = sizeof(Pair<T, int>) * kBlockMinCount;
         auto grid =
                 dim3(utility::DivUp(outDistances.GetShape(0), kRowsPerBlock));
 
@@ -230,7 +234,7 @@ void runL2SelectMin(const cudaStream_t stream,
         int64_t out_row_stride = outDistances.GetStride(0);
 
         l2SelectMin1<T, TIndex, kRowsPerBlock, kThreadsPerBlock>
-                <<<grid, block, 0, stream>>>(productDistances.GetDataPtr<T>(),
+                <<<grid, block, shared_mem_size, stream>>>(productDistances.GetDataPtr<T>(),
                                              centroidDistances.GetDataPtr<T>(),
                                              outDistances.GetDataPtr<T>(),
                                              outIndices.GetDataPtr<TIndex>(),
